GPU/Vulkan/StencilBufferVulkan.cpp

```diff
@@ -0,0 +1,219 @@
+#extension GL_ARB_shading_language_420pack : enable
+layout (location = 0) out vec2 v_texcoord0;
+out gl_PerVertex { vec4 gl_Position; };
+void main() {
+	int id = gl_VertexIndex;
+  v_texcoord0.x = (id == 2) ? 2.0 : 0.0;
+  v_texcoord0.y = (id == 1) ? 2.0 : 0.0;
+  gl_Position = vec4(v_texcoord0 * vec2(2.0, 2.0) + vec2(-1.0, -1.0), 0.0, 1.0);
+}
+)";
+
+static u8 StencilBits5551(const u8 *ptr8, u32 numPixels) {
+	const u32 *ptr = (const u32 *)ptr8;
+
+	for (u32 i = 0; i < numPixels / 2; ++i) {
+		if (ptr[i] & 0x80008000) {
+			return 1;
+		}
+	}
+	return 0;
+}
+
+static u8 StencilBits4444(const u8 *ptr8, u32 numPixels) {
+	const u32 *ptr = (const u32 *)ptr8;
+	u32 bits = 0;
+
+	for (u32 i = 0; i < numPixels / 2; ++i) {
+		bits |= ptr[i];
+	}
+
+	return ((bits >> 12) & 0xF) | (bits >> 28);
+}
+
+static u8 StencilBits8888(const u8 *ptr8, u32 numPixels) {
+	const u32 *ptr = (const u32 *)ptr8;
+	u32 bits = 0;
+
+	for (u32 i = 0; i < numPixels; ++i) {
+		bits |= ptr[i];
+	}
+
+	return bits >> 24;
+}
+
+// In Vulkan we should be able to simply copy the stencil data directly to a stencil buffer without
+// messing about with bitplane textures and the like. Or actually, maybe not... Let's start with
+// the traditional approach.
+bool FramebufferManagerVulkan::NotifyStencilUpload(u32 addr, int size, bool skipZero) {
+	if (!MayIntersectFramebuffer(addr)) {
+		return false;
+	}
+
+	VirtualFramebuffer *dstBuffer = 0;
+	for (size_t i = 0; i < vfbs_.size(); ++i) {
+		VirtualFramebuffer *vfb = vfbs_[i];
+		if (MaskedEqual(vfb->fb_address, addr)) {
+			dstBuffer = vfb;
+		}
+	}
+	if (!dstBuffer) {
+		return false;
+	}
+
+	int values = 0;
+	u8 usedBits = 0;
+
+	const u8 *src = Memory::GetPointer(addr);
+	if (!src) {
+		return false;
+	}
+
+	switch (dstBuffer->format) {
+	case GE_FORMAT_565:
+		// Well, this doesn't make much sense.
+		return false;
+	case GE_FORMAT_5551:
+		usedBits = StencilBits5551(src, dstBuffer->fb_stride * dstBuffer->bufferHeight);
+		values = 2;
+		break;
+	case GE_FORMAT_4444:
+		usedBits = StencilBits4444(src, dstBuffer->fb_stride * dstBuffer->bufferHeight);
+		values = 16;
+		break;
+	case GE_FORMAT_8888:
+		usedBits = StencilBits8888(src, dstBuffer->fb_stride * dstBuffer->bufferHeight);
+		values = 256;
+		break;
+	case GE_FORMAT_INVALID:
+		// Impossible.
+		break;
+	}
+
+	std::string error;
+	if (!stencilVs_) {
+		stencilVs_ = CompileShaderModule(vulkan_, VK_SHADER_STAGE_VERTEX_BIT, stencil_vs, &error);
+		stencilFs_ = CompileShaderModule(vulkan_, VK_SHADER_STAGE_FRAGMENT_BIT, stencil_fs, &error);
+	}
+	VkRenderPass rp = (VkRenderPass)draw_->GetNativeObject(Draw::NativeObject::FRAMEBUFFER_RENDERPASS);
+
+	VulkanRenderManager *renderManager = (VulkanRenderManager *)draw_->GetNativeObject(Draw::NativeObject::RENDER_MANAGER);
+
+	if (usedBits == 0) {
+		if (skipZero) {
+			// Common when creating buffers, it's already 0.  We're done.
+			return false;
+		}
+
+		// TODO: Find a nice way to clear alpha here too.
```

Comment:
VulkanRenderManager::Clear?  I know it's ideally avoided, but, here we want to just clear one channel so...

-[Unknown]

Reply:
Clear can't clear just alpha. RGBA channel masks don't actually work on clears on a lot of hardware and are not present for clears in Vulkan, I assume drivers tend to emulate glClear with masks with a draw call. 

So yeah, I'm gonna have to do a draw call here.

Reply:
Ah right, yes.

-[Unknown]